fortuna/ood_detection/ddu.py

```diff
@@ -0,0 +1,193 @@
+    of the feature representation.
+
+    See `Mukhoti, Jishnu, et al. <https://arxiv.org/abs/2102.11582>`_
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(DeepDeterministicUncertaintyABC, self).__init__(*args, **kwargs)
+        self._gmm_logpdf_fn = None
+
+    def fit(
+        self,
+        state: PosteriorState,
+        train_data_loader: BaseDataLoaderABC,
+        num_classes: int,
+    ) -> None:
+        """
+        Fits a Multivariate Gaussian to the training data using class-specific means and covariance matrix.
+
+        Parameters
+        ----------
+        state: PosteriorState
+            the posterior state ob a pre-trained model
+        train_data_loader: BaseDataLoaderABC
+            the training data loader (covariates and target)
+        num_classes: int
+            the number of classes for the training task
+        """
+        train_labels = []
+        train_embeddings = []
+        for x, y in tqdm(train_data_loader, desc="Computing embeddings for DDU: "):
+            train_embeddings.append(
+                self.apply(inputs=x, params=state.params, mutable=state.mutable)
+            )
+            train_labels.append(y)
+        train_embeddings = jnp.concatenate(train_embeddings, 0)
+        train_labels = jnp.concatenate(train_labels)
+
+        (
+            classwise_mean_features,
+            classwise_cov_features,
+        ) = compute_classwise_mean_and_cov(train_embeddings, train_labels, num_classes)
+        self._gmm_logpdf_fn = _get_logpdf_fn(
+            classwise_mean_features, classwise_cov_features
+        )
+
+    def score(self, state: PosteriorState, inputs_loader: BaseInputsLoader) -> Array:
+        """
+        The confidence score :math:`M(\mathbf{x})` for a new test sample :math:`\mathbf{x}` is obtained computing
+        the negative marginal likelihood of the feature representation
+        :math:`-q(f(\mathbf{x})) = - \sum\limits_{k}q(f(\mathbf{x})|y) q(y)`.
+
+        A high score signals that the test sample :math:`\mathbf{x}` is identified as OOD.
+
+        Parameters
+        ----------
+        state: PosteriorState
+            The posterior state of a pre-trained model
+        inputs_loader:  BaseInputsLoader
+            The inputs loader (data only, no labels)
+
+        Returns
+        -------
+        Array
+            An array with scores for each sample in `inputs_loader`.
+        """
+        if self._gmm_logpdf_fn is None:
+            raise NotFittedError("You have to call fit before calling score.")
+        embeddings = jnp.concatenate(
+            [
+                self.apply(inputs=x, params=state.params, mutable=state.mutable)
+                for x in inputs_loader
+            ],
+            0,
+        )
+        loglik = self._gmm_logpdf_fn(embeddings)
+        return -jsp.special.logsumexp(jnp.nan_to_num(loglik, 0.0), axis=1)
```

Comment:
I don't get why the confidence has the minus in front. It looks like you by marginalizing you are computing a positive quantity. Shouldn't also the score be positive?

Also, should one make sure that the score is between 0 and 1?